### DISCUSSION TITLE
Trigger an action after proposing block

### DISCUSSION THREAD

**Discussion by guilherme-brandao** [cosmos/community#74] [Chain Builders (Cosmos SDK)] [answered]
I would like to trigger a specific action after the leader proposes the block; only the leader needs to do it.

I was searching for something similar to `EndBlock`, but again, just for the validator that is proposing the block.

**Comment by facundomedica** [comment, marked as answer]
Hey! I think what you are referring to can be done in PrepareProposal, you can add a custom message in the block that will get executed. You'll also give a chance to the other validators to verify that this custom message is actually valid in ProcessProposal.

**Reply by guilherme-brandao** [reply]
Can I add it inside a custom module? The same way that I add the `EndBlock`? How can I enable it?

**Reply by facundomedica** [reply]
No, it should be defined at app level (app.go). See this tutorial for more in-depth explanations and examples: https://github.com/cosmos/sdk-tutorials/blob/master/tutorials/oracle/docs/00-getting-started.md

You'll see that besides the PrepareProposal and ProcessProposal handlers we also define vote extensions, I think you won't need those for now
